src/Microsoft.Diagnostics.Monitoring.EventPipe/Counters/CounterPayload.cs

```diff
@@ -86,7 +86,38 @@ protected MeterPayload(DateTime timestamp,
         {
         }
 
-        public override bool IsMeter => true;
+        public sealed override bool IsMeter => true;
+    }
+
+    internal abstract class MeterInstrumentDeltaMeasurementPayload : MeterPayload
+    {
+        protected MeterInstrumentDeltaMeasurementPayload(
+            DateTime timestamp,
+            CounterMetadata counterMetadata,
+            string displayName,
+            string unit,
+            double value,
+            CounterType counterType,
+            string valueTags,
+            EventType eventType)
+            : base(timestamp, counterMetadata, displayName, unit, value, counterType, valueTags, eventType)
+        {
+        }
+
+        public abstract bool SupportsDelta { get; }
+
+        public virtual void Combine(MeterInstrumentDeltaMeasurementPayload other)
```

Comment:
Right now all these payload objects are simple immutable objects that represent data which came across the wire and it would be nice to preserve that immutability. Rather than modify the objects in place perhaps you could track the mutating state in a couple variables you define together with the for loop iterating over them? That seems simpler and clearer to me than trying to reuse the payload object also as an accumulator for different subsets of the properties.

Reply:
I dropped the combine concept. I do like the immutability of these objects too. They can mutate DisplayName it seems but still 😄 Hard to know _exactly_ how the aggregation will work. I took a look at my PoC. At the moment it passes a payload type. So it will have to allocate a new one with the final value. But that may not be the end of the world, or something else could be possible, or we could do more here. But figured OK to get the good pieces in so it can keep moving/evolving.